src/Runner.Worker/JobExtension.cs

```diff
@@ -365,55 +365,83 @@ public void FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRequestM
-                                    context.Warning($"Skip output '{output.Key}' since it may contain secret.");
-                                    continue;
-                                }
+                                    if (string.IsNullOrEmpty(output.Value))
+                                    {
+                                        context.Debug($"Skip output '{output.Key}' since it's empty");
+                                        continue;
+                                    }
 
-                                context.Output($"Set output '{output.Key}'");
-                                jobContext.JobOutputs[output.Key] = output.Value;
+                                    if (!string.Equals(output.Value, HostContext.SecretMasker.MaskSecrets(output.Value)))
+                                    {
+                                        context.Warning($"Skip output '{output.Key}' since it may contain secret.");
+                                        continue;
+                                    }
+
+                                    context.Output($"Set output '{output.Key}'");
+                                    jobContext.JobOutputs[output.Key] = output.Value;
+                                }
+                            }
+                            catch (Exception ex)
+                            {
+                                context.Result = TaskResult.Failed;
+                                context.Error($"Fail to evaluate job outputs");
+                                context.Error(ex);
+                                jobContext.Result = TaskResultUtil.MergeTaskResults(jobContext.Result, TaskResult.Failed);
                             }
                         }
-                        catch (Exception ex)
+
+                        // Evaluate environment data
+                        if (evaluateEnvironmentData)
                         {
-                            context.Result = TaskResult.Failed;
-                            context.Error($"Fail to evaluate job outputs");
-                            context.Error(ex);
-                            jobContext.Result = TaskResultUtil.MergeTaskResults(jobContext.Result, TaskResult.Failed);
+                            try
+                            {
+                                context.Output($"Evaluate and set environment url");
+
+                                // Populate env context for each step
+                                Trace.Info("Initialize Env context for evaluating environment url");
+                                var environmentUrlToken = templateEvaluator.EvaluateEnvironmentUrl(jobContext.ActionsEnvironment.Url, context.ExpressionValues, context.ExpressionFunctions);
+                                context.Output($"Evaluated environment url: {environmentUrlToken.AssertString("environment.url")}");
+                                jobContext.ActionsEnvironment.Url = environmentUrlToken;
```

Comment:
scrub secrets (for example .env or step outputs may have secrets)

Reply:
see `if (!string.Equals(output.Value, HostContext.SecretMasker.MaskSecrets(output.Value)))` above